src/policy/vmspace.rs

```diff
@@ -112,118 +144,104 @@ impl<VM: VMBinding> crate::policy::gc_work::PolicyTraceObject<VM> for VMSpace<VM
 }
 
 impl<VM: VMBinding> VMSpace<VM> {
-    pub fn new(args: &mut CreateSpecificPlanArgs<VM>) -> Self {
-        let args_clone = CreateSpecificPlanArgs {
-            global_args: CreateGeneralPlanArgs {
-                vm_map: args.global_args.vm_map,
-                mmapper: args.global_args.mmapper,
-                heap: HeapMeta::new(), // we do not use this
-                options: args.global_args.options.clone(),
-                gc_trigger: args.global_args.gc_trigger.clone(),
-                scheduler: args.global_args.scheduler.clone(),
-            },
-            constraints: args.constraints,
-            global_side_metadata_specs: args.global_side_metadata_specs.clone(),
+    pub fn new(args: crate::policy::space::PlanCreateSpaceArgs<VM>) -> Self {
+        let (vm_space_start, vm_space_size) =
+            (*args.options.vm_space_start, *args.options.vm_space_size);
+        let space = Self {
+            mark_state: MarkState::new(),
+            pr: ExternalPageResource::new(args.vm_map),
+            common: CommonSpace::new(args.into_policy_args(
+                false,
+                true,
+                crate::util::metadata::extract_side_metadata(&[
+                    *VM::VMObjectModel::LOCAL_MARK_BIT_SPEC,
+                ]),
+            )),
         };
-        // Create the space if the VM space start/size is set. Otherwise, use None.
-        let inner = (!args.global_args.options.vm_space_start.is_zero())
-            .then(|| Self::create_space(args, None));
-        Self {
-            inner,
-            args: args_clone,
+
+        if !vm_space_start.is_zero() {
+            space.add_external_pages(vm_space_start, vm_space_size);
         }
-    }
 
-    pub fn lazy_initialize(&mut self, start: Address, size: usize) {
-        assert!(self.inner.is_none(), "VM space has been initialized");
-        self.inner = Some(Self::create_space(&mut self.args, Some((start, size))));
+        space
+    }
 
-        self.common().initialize_sft(self.as_sft());
+    pub fn set_vm_region(&mut self, start: Address, size: usize) {
+        self.add_external_pages(start, size);
     }
 
-    fn create_space(
-        args: &mut CreateSpecificPlanArgs<VM>,
-        location: Option<(Address, usize)>,
-    ) -> ImmortalSpace<VM> {
-        use crate::util::heap::layout::vm_layout_constants::BYTES_IN_CHUNK;
+    pub fn add_external_pages(&self, start: Address, size: usize) {
+        let start = start.align_down(BYTES_IN_PAGE);
+        let end = (start + size).align_up(BYTES_IN_PAGE);
+        let size = end - start;
```

Comment:
When we calculate `end`, `start` has already been overwritten.  If size is a multiple of pages, `end` will not include the last page the original `start` and `size` covered.

Because there are variables named `chunk_start` and `chunk_end` below, I suggest prefixing the computed values with `page_`.

```suggestion
    pub fn add_external_pages(&self, start: Address, start: usize) {
        let page_start = start.align_down(BYTES_IN_PAGE);
        let page_end = (start + start).align_up(BYTES_IN_PAGE);
        let page_size = page_end - page_start;
```